vumi/persist/tests/test_fake_redis.py

```diff
@@ -1,491 +1,750 @@
-        yield self.assert_redis_op('list', 'type', 'list_key')
-        yield self.redis.sadd("set_key", "a")
-        yield self.assert_redis_op('set', 'type', 'set_key')
-        yield self.redis.zadd("zset_key", a=1.0)
-        yield self.assert_redis_op('zset', 'type', 'zset_key')
-        yield self.redis.hset("hash_key", "a", 1.0)
-        yield self.assert_redis_op('hash', 'type', 'hash_key')
+        redis = yield self.get_redis()
+        yield self.assert_redis_op(redis, 'none', 'type', 'unknown_key')
+        yield redis.set("string_key", "a")
+        yield self.assert_redis_op(redis, 'string', 'type', 'string_key')
+        yield redis.lpush("list_key", "a")
+        yield self.assert_redis_op(redis, 'list', 'type', 'list_key')
+        yield redis.sadd("set_key", "a")
+        yield self.assert_redis_op(redis, 'set', 'type', 'set_key')
+        yield redis.zadd("zset_key", a=1.0)
+        yield self.assert_redis_op(redis, 'zset', 'type', 'zset_key')
+        yield redis.hset("hash_key", "a", 1.0)
+        yield self.assert_redis_op(redis, 'hash', 'type', 'hash_key')
+
+    @inlineCallbacks
+    def test_charset_encoding_default(self):
+        # Redis client assumes utf-8
+        redis = yield self.get_redis()
+        yield redis.set('name', u'Zoë Destroyer of Ascii')
+        yield self.assert_redis_op(
+            redis, 'Zo\xc3\xab Destroyer of Ascii', 'get', 'name')
+
+    @inlineCallbacks
+    def test_charset_encoding_custom_replace(self):
+        redis = yield self.get_redis(charset='ascii', errors='replace')
+        yield redis.set('name', u'Zoë Destroyer of Ascii')
+        yield self.assert_redis_op(
+            redis, 'Zo? Destroyer of Ascii', 'get', 'name')
+
+    @inlineCallbacks
+    def test_charset_encoding_custom_ignore(self):
+        redis = yield self.get_redis(charset='ascii', errors='ignore')
+        yield redis.set('name', u'Zoë Destroyer of Ascii')
+        yield self.assert_redis_op(
+            redis, 'Zo Destroyer of Ascii', 'get', 'name')
+
+    @inlineCallbacks
+    def test_scan_no_keys(self):
+        """
+        Real and fake Redis implementation all return the same (empty) response
+        when we scan for keys that don't exist. Other scanning methods are in
+        FakeRedisUnverifiedTestMixin because we can't fake the same arbitrary
+        order in which keys are returned from real Redis.
+        """
+        redis = yield self.get_redis()
+        yield self.assert_redis_op(redis, [None, []], 'scan', None)
+
+
+class FakeRedisUnverifiedTestMixin(object):
+    """
+    This mixin adds some extra tests that are not verified against real Redis.
+
+    Each test in here should explain why verification isn't possible.
+    """
 
     @inlineCallbacks
     def test_scan_simple(self):
+        """
+        Scanning returns keys in an order that depends on arbitrary state in
+        the Redis server, so we can't fake it in a way that's identical to real
+        Redis.
+        """
```

Comment:
Yeah. I tried to rewrite the tests in a way that would let us walk the scan over the whole set of keys and check that we got the same set of results at the end, but that ended up being more effort than it's really worth.